dali/pipeline/operators/reader/nvdecoder/nvdecoder.cc

```diff
@@ -142,42 +145,54 @@ int NvDecoder::handle_display(void* user_data,
   return decoder->handle_display_(disp_info);
 }
 
+// Prepare params and calls cuvidCreateDecoder
 int NvDecoder::handle_sequence_(CUVIDEOFORMAT* format) {
   frame_base_ = {static_cast<int>(format->frame_rate.denominator),
                   static_cast<int>(format->frame_rate.numerator)};
-
-  // Prepare params and calls cuvidCreateDecoder
-  return decoder_.initialize(format);
+  int ret = kNvcuvid_failure;
+  try {
+    ret = decoder_.initialize(format);
+  } catch (...) {
+    done_ = true;
+    captured_exception_ = std::current_exception();
+    // Main thread is waiting on frame_queue_
+    frame_queue_.shutdown();
+  }
+  return ret;
 }
 
 int NvDecoder::handle_decode_(CUVIDPICPARAMS* pic_params) {
   int total_wait = 0;
   constexpr auto sleep_period = 500;
   constexpr auto timeout_sec = 20;
   constexpr auto enable_timeout = false;
+
+  // If something went wrong during init we exit directly
+  if (done_) return kNvcuvid_failure;
+
   while (frame_in_use_[pic_params->CurrPicIdx]) {
     if (enable_timeout &&
       total_wait++ > timeout_sec * 1000000 / sleep_period) {
-      std::cout << device_id_ << ": Waiting for picture "
-                << pic_params->CurrPicIdx
-                << " to become available..." << std::endl;
+      LOG_LINE << device_id_ << ": Waiting for picture "
+               << pic_params->CurrPicIdx
+               << " to become available..." << std::endl;
       std::stringstream ss;
       ss << "Waited too long (" << timeout_sec << " seconds) "
-          << "for decode output buffer to become available";
+         << "for decode output buffer to become available";
       DALI_FAIL(ss.str());
     }
     usleep(sleep_period);
-    if (done_) return 0;
+    if (done_) return kNvcuvid_failure;
   }
 
   LOG_LINE << "Sending a picture for decode"
-              << " size: " << pic_params->nBitstreamDataLen
-              << " pic index: " << pic_params->CurrPicIdx
-              << std::endl;
+           << " size: " << pic_params->nBitstreamDataLen
+           << " pic index: " << pic_params->CurrPicIdx
+           << std::endl;
 
   // decoder_ operator () returns a CUvideodecoder
   CUDA_CALL(cuvidDecodePicture(decoder_, pic_params));
-  return 1;
+  return kNvcuvid_success;
```

Comment:
Where we consume this value? Do we use kNvcuvid_success/kNvcuvid_failure there as well?

Reply:
There are the returns of callback which value is interpreted by the NVDEC decoder

Reply:
So I guess we need to use this enum/define/constexp values there as well.

Reply:
There is no enum specified in the docs or here https://github.com/NVIDIA/DALI/blob/19f8d2dd8eab7c71b5c50f74c1d4d0f5a91c6e8a/dali/pipeline/operators/reader/nvdecoder/dynlink_nvcuvid.h#L302 . I think the return value is just interpreted as some kind boolean. Maybe we can use booleans casted to int?